post-repo-inmem/src/test/java/klon/post/repo/inmem/PostRepositoryInmemImplTest.java

```diff
@@ -0,0 +1,118 @@
+import static org.awaitility.Awaitility.await;
+import static org.junit.jupiter.api.Assertions.*;
+
+class PostRepositoryInmemImplTest {
+
+    PostRepositoryInmemImpl repo = null;
+
+    @BeforeEach
+    void setUp() {
+        repo = new PostRepositoryInmemImpl(PostMapper.INSTANCE);
+    }
+
+    @Test
+    void shouldStorePost() {
+        Post post = Post.builder()
+                .postId(randomAlphanumeric(10))
+                .content(randomAlphanumeric(100))
+                .userId(randomAlphanumeric(10))
+                .build();
+        repo.addPost(post);
+        Post fromStore = repo.getPost(post.getPostId()).get();
+
+        assertEquals(post.getPostId(), fromStore.getPostId());
+        assertEquals(post.getContent(), fromStore.getContent());
+        assertEquals(post.getCreatedTime(), fromStore.getCreatedTime());
+        assertEquals(post.getUserId(), fromStore.getUserId());
+    }
+
+    @Test
+    void shouldAddPostToCollectionForUser() {
+        Post older = Post.builder()
+                .postId(randomAlphanumeric(10))
+                .userId(randomAlphanumeric(10))
+                .build();
+        Post newer = Post.builder()
+                .postId(randomAlphanumeric(10))
+                .userId(older.getUserId())
+                .build();
+        repo.addPost(older);
+        repo.addPost(newer);
+        Stream<Post> postsByUser = repo.getPostsByUser(older.getUserId());
+
+        assertEquals(2, postsByUser.count());
+    }
+
+    @Test
+    void shouldReturnEmptyOptionalIfPostNotExist() {
+        Optional<Post> post = repo.getPost(randomAlphanumeric(10));
+
+        assertTrue(post.isEmpty());
+    }
+
+    @Test
+    void shouldReturnNonEmptyOptionalForExistingPost() {
+        Post post = Post.builder().userId(randomAlphanumeric(10)).postId(randomAlphanumeric(10)).build();
+        repo.addPost(post);
+        Optional<Post> fromStore = repo.getPost(post.getPostId());
+
+        assertFalse(fromStore.isEmpty());
+    }
+
+    @Test
+    void shouldReturnEmptyCollectionForNonExistingUser() {
+        Stream<Post> postsByUser = repo.getPostsByUser(randomAlphanumeric(10));
+
+        assertEquals(0, postsByUser.count());
+    }
+
+    @Test
+    void shouldReturnStreamStartingWithPostIdDroppingOlderPosts() {
+        Post older = Post.builder().userId(randomAlphanumeric(10)).postId(randomAlphanumeric(10)).createdTime(ZonedDateTime.now()).build();
+        Post newer = Post.builder().userId(older.getUserId()).postId(randomAlphanumeric(10)).createdTime(ZonedDateTime.now()).build();
+
+        repo.addPost(older);
+        repo.addPost(newer);
+
+        Stream<Post> postsByUser = repo.getPostsByUser(older.getUserId(), newer.getPostId());
+        assertEquals(1, postsByUser.count());
+    }
+
+    @Test
+    @SneakyThrows
+    void addNewPostInAnotherThreadShouldNotAffectReadingPosts() {
+        CountDownLatch latch = new CountDownLatch(2);
```

Comment:
Nie znam Awaitility, ale czy pollDelay gwarantuje, że wątek z 1.5 sec opóźnienia zadziała po wątku z 1sec, jeśli na mocno obciążonym środowisku będzie brakowało zasobów i oba zawisną na 2s?
Może dodatkowe CountdownLatch(e) zamiast pollDelay?

Reply:
Też tego nie znam. Sonar to zaproponował zamiast użycia Thread.sleep(), które miało u mnie symulować opóźnienie. Nie ma problemu, użyje dodatkowego CountDownLatcha

Reply:
Tak odnośnie _Thread.sleep_, to wygodniejsze jest [_TimeUnit.sleep_](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/TimeUnit.html#sleep(long))